src/user-data.ts

```diff
@@ -131,3 +130,196 @@ export type AppSemester = {
+        // Add section
+        const enroll = section.ssrComponent;
+        enrollmentMap[enroll] = true;
+
+        section.meetings.forEach(meeting => {
+          const { pattern, timeStart, timeEnd } = meeting;
+          // Only add the time if it is a lecture
+          if (enroll === 'LEC') lectureTimesMap[`${pattern} ${timeStart} - ${timeEnd}`] = true;
+
+          meeting.instructors.forEach(instructor => {
+            const { netid, firstName, lastName } = instructor;
+            instructorsMap[netid] = `${firstName} ${lastName}`;
+          });
+        });
+      });
+    });
+
+    enrollment = Object.keys(enrollmentMap);
+    lectureTimes = Object.keys(lectureTimesMap);
+    instructors = Object.keys(instructorsMap).map(netid => `${instructorsMap[netid]} (${netid})`);
+  }
+
+  // Distribution of course (e.g. MQR-AS)
+  // alternateDistributions option in case catalogDistr for the course is null, undef, ''
+  const catalogDistrDoesNotExist = !course.catalogDistr || course.catalogDistr === '';
+  const alternateDistributions = catalogDistrDoesNotExist
+    ? ['']
+    : /\(([^)]+)\)/.exec(course.catalogDistr!)![1].split(', ');
+  const distributions = course.distributions || alternateDistributions;
+
+  // Get last semester of available course. TODO: Remove when no longer firebase data dependant
+  const lastRoster = course.lastRoster || course.roster;
+
+  // Create course from saved color. Otherwise, create course from subject color group
+  const color = course.color || addColor(subject);
+
+  const alerts = { requirement: null, caution: null };
+
+  return {
+    crseId: course.crseId,
+    subject,
+    number,
+    name,
+    description,
+    credits,
+    creditRange,
+    semesters,
+    prereqs,
+    enrollment,
+    lectureTimes,
+    instructors,
+    distributions,
+    lastRoster,
+    color,
+    alerts,
+    check: true,
+    uniqueID,
+  };
+};
+
+export const firestoreSemesterToAppSemester = (
+  { courses, type, year }: FirestoreSemester,
+  semesterID: number,
+  incrementID: () => number,
+  addColor: (subject: string) => string
+): AppSemester => {
+  return {
+    id: semesterID,
+    courses: courses.map(course => firestoreCourseToAppCourse(course, incrementID, addColor)),
+    type,
+    year,
+  };
+};
+
+export const createAppUser = (data: FirestoreNestedUserData, name: FirestoreUserName): AppUser => {
+  const user: AppUser = {
+    // TODO: take into account multiple majors and colleges
+    college: data.colleges[0].acronym,
+    collegeFN: data.colleges[0].fullName,
+    firstName: name.firstName,
+    middleName: name.middleName,
+    lastName: name.lastName,
+    major: [],
+    majorFN: [],
+    minor: [],
+    minorFN: [],
+    exam: [],
+    transferCourse: [],
+    tookSwim: data.tookSwim,
+  };
+  if ('exam' in data && data.exam.length > 0) {
+    user.exam = [...data.exam];
+  }
+  if ('class' in data && data.class.length > 0) {
+    user.transferCourse = [...data.class];
+  }
```

Comment:
In general, I really don't like this kind of field existence conditional check. We should ideally write scripts to migrate all the old data, instead of using this kind of undocumented hack.